server/lizmap_service.py

```diff
@@ -0,0 +1,420 @@
+                if l.type() == QgsMapLayer.VectorLayer:
+                    layer = l
+                    break
+        # find with short name
+        if not layer:
+            maplayers = project.mapLayersByShortName(layername)
+            if maplayers:
+                for l in maplayers:
+                    if l.type() == QgsMapLayer.VectorLayer:
+                        layer = l
+                        break
+        # find with layer id
+        if not layer:
+            layer = project.mapLayer(layername)
+        # layer not found
+        if not layer:
+            raise ExpressionServiceError(
+                "Bad request error",
+                "Invalid LAYER parameter for 'EVALUATE': {} provided".format(layername),
+                400)
+
+        # get expressions
+        expressions = params.get('EXPRESSIONS', '')
+        if not expressions:
+            expression = params.get('EXPRESSION', '')
+            if not expression:
+                raise ExpressionServiceError(
+                    "Bad request error",
+                    "Invalid 'EVALUATE' REQUEST: EXPRESSION or EXPRESSIONS parameter is mandatory",
+                    400)
+            expressions = '["{}"]'.format(expression)
+
+        # try to load expressions list or dict
+        exp_json = None
+        try:
+            exp_json = json.loads(expressions)
+        except Exception:
+            QgsMessageLog.logMessage(
+                "JSON loads expressions '{}' exception:\n{}".format(expressions, traceback.format_exc()),
+                "lizmap", Qgis.Critical)
+            raise ExpressionServiceError(
+                "Bad request error",
+                "Invalid 'EVALUATE' REQUEST: EXPRESSIONS '{}' are not well formed".format(data),
+                400)
+
+        # create expression context
+        exp_context = QgsExpressionContext()
+        exp_context.appendScope(QgsExpressionContextUtils.globalScope())
+        exp_context.appendScope(QgsExpressionContextUtils.projectScope(project))
+        exp_context.appendScope(QgsExpressionContextUtils.layerScope(layer))
+
+        # create distance area context
+        da = QgsDistanceArea()
+        da.setSourceCrs(layer.crs(), project.transformContext())
+        da.setEllipsoid(project.ellipsoid())
+
+        # parse expressions
+        exp_map = {}
+        exp_parser_errors = []
+        exp_items = []
+        if isinstance(exp_json, list):
+            exp_items = enumerate(exp_json)
+        elif isinstance(exp_json, dict):
+            exp_items = exp_json.items()
+        for k, e in exp_items:
+            exp = QgsExpression(e)
+            exp.setGeomCalculator(da)
+            exp.setDistanceUnits(project.distanceUnits())
+            exp.setAreaUnits(project.areaUnits())
+
+            if exp.hasParserError():
+                exp_parser_errors.append('Error "{}": {}'.format(e, exp.parserErrorString()))
+                continue
+
+            if not exp.isValid():
```

Comment:
`hasEvalError`?

Reply:
not yet, the expression as not been yet evaluated.